### PR TITLE
[core] Prevent internal dev-properties from being displayed on CodeClimate renderer

### LINKED ISSUE #643: [core] PMD Properties (dev-properties) breaks markup on CodeClimateRenderer
**Rule Set:** java-bsic

**Description:** Some rules generate an XPath snippet that looks weird in the **bottom** of the report

**Code Sample demonstrating the issue:**
```java
public class Library {
	public void foo() {
		for (int i = 0; i < 10; i++) {
			for (int k = 0; k < 20; i++) {
				System.out.println("Hello");
			}
		}
	}
}
```

**Running PMD through:** `./run.sh pmd -d . -f codeclimate -R java-basic`
```js
{
  "type": "issue",
  "check_name": "JumbledIncrementer",
  "description": "Avoid modifying an outer loop incrementer in an inner loop for update expression",
  "content": {
    "body": "## JumbledIncrementer\n\nSince: PMD 1.0\n\nPriority: Medium\n\n[Categories](https://github.com/codeclimate/spec/blob/master/SPEC.md#categories): Style\n\n[Remediation Points](https://github.com/codeclimate/spec/blob/master/SPEC.md#remediation-points): 50000\n\nAvoid jumbled loop incrementers - its usually a mistake, and is confusing even if intentional.\n\n### Example:\n\n```java\n\n \npublic class JumbledIncrementerRule1 {\n\tpublic void foo() {\n\t\tfor (int i = 0; i < 10; i++) {\t\t\t// only references 'i'\n\t\t\tfor (int k = 0; k < 20; i++) {\t\t// references both 'i' and 'k'\n\t\t\t\tSystem.out.println('Hello');\n\t\t\t}\n\t\t}\n\t}\n}\n \n \n``` \n\n### [PMD properties](http://pmd.github.io/pmd-5.8.1/customizing/pmd-developer.html)\n\nName | Value | Description\n--- | --- | ---\nviolationSuppressRegex | | Suppress violations with messages matching a regular expression\nviolationSuppressXPath | | Suppress violations on nodes which match a given relative XPath expression.\nversion | 1.0 | XPath specification version\nxpath | \n \n//ForStatement\n [\n ForUpdate/StatementExpressionList/StatementExpression/PostfixExpression/PrimaryExpression/PrimaryPrefix/Name/@Image\n =\n ancestor::ForStatement/ForInit//VariableDeclaratorId/@Image\n ]\n \n | XPath expression\n"
  },
  "categories": [
    "Style"
  ],
  "location": {
    "path": "/Users/filipe/codeclimate/tmp/java/simple_java/src/main/java/Library.java",
    "lines": {
      "begin": 4,
      "end": 6
    }
  },
  "severity": "normal",
  "remediation_points": 50000
}
```

## JumbledIncrementer

Since: PMD 1.0

Priority: Medium

[Categories](https://github.com/codeclimate/spec/blob/master/SPEC.md#categories): Style

[Remediation Points](https://github.com/codeclimate/spec/blob/master/SPEC.md#remediation-points): 50000

Avoid jumbled loop incrementers - its usually a mistake, and is confusing even if intentional.

### Example:

```java

 
public class JumbledIncrementerRule1 {
\tpublic void foo() {
\t\tfor (int i = 0; i < 10; i++) {\t\t\t// only references 'i'
\t\t\tfor (int k = 0; k < 20; i++) {\t\t// references both 'i' and 'k'
\t\t\t\tSystem.out.println('Hello');
\t\t\t}
\t\t}
\t}
}
 
 
``` 

### [PMD properties](http://pmd.github.io/pmd-5.8.1/customizing/pmd-developer.html)

Name | Value | Description
--- | --- | ---
violationSuppressRegex | | Suppress violations with messages matching a regular expression
violationSuppressXPath | | Suppress violations on nodes which match a given relative XPath expression.
version | 1.0 | XPath specification version
xpath | 
 
//ForStatement
 [
 ForUpdate/StatementExpressionList/StatementExpression/PostfixExpression/PrimaryExpression/PrimaryPrefix/Name/@Image
 =
 ancestor::ForStatement/ForInit//VariableDeclaratorId/@Image
 ]
 
 | XPath expression

### PR DESCRIPTION
**Pull request by filipesperandio**
… documentation markdown

Fix #643

Please, prefix the PR title with the language it applies to within brackets, such as *[java]* or *[apex]*. If not specific to a language, you can use *[core]*

Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `mvn test` passes.
 - [x] `mvn checkstyle:check` passes. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/CodeClimateRenderer.java [hunk 1]
```diff
@@ -35,6 +35,7 @@
 
     // Note: required by https://github.com/codeclimate/spec/blob/master/SPEC.md
     protected static final String NULL_CHARACTER = "\u0000";
+    protected static final List<String> INTERNAL_DEV_PROPERTIES = Arrays.asList("index", "xpath");
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/renderers/CodeClimateRenderer.java]: should be `version` and `xpath`

Comment (filipesperandio) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/CodeClimateRenderer.java]: :facepalm: Not sure where `index` came from! haha

**Comment by filipesperandio** [comment]
@jsotuyod Having trouble to manually test it, though.
I can manage to get the `.zip`, extract it into a `/tmp` folder, but not seeing the change in the behaviour. Feels like I am missing something.

**Comment by filipesperandio** [comment]
@jsotuyod I got it.

This is the output:

## JumbledIncrementer

Since: PMD 1.0

Priority: Medium

[Categories](https://github.com/codeclimate/spec/blob/master/SPEC.md#categories): Style

[Remediation Points](https://github.com/codeclimate/spec/blob/master/SPEC.md#remediation-points): 50000

Avoid jumbled loop incrementers - its usually a mistake, and is confusing even if intentional.

### Example:

```java

 
public class JumbledIncrementerRule1 {
 public void foo() {
 for (int i = 0; i < 10; i++) { // only references 'i'
 for (int k = 0; k < 20; i++) { // references both 'i' and 'k'
 System.out.println('Hello');
 }
 }
 }
}

 
``` 

### [PMD properties](http://pmd.sourceforge.net/snapshot/customizing/pmd-developer.html)

Name | Value | Description
--- | --- | ---
violationSuppressRegex | | Suppress violations with messages matching a regular expression
violationSuppressXPath | | Suppress violations on nodes which match a given relative XPath expression.

**Comment by jsotuyod** [comment]
@filipesperandio cool! I'll add a JUnit test myself just in case, but that's great!

**Comment by filipesperandio** [comment]
@jsotuyod What is the timeline/process for releases?
When will this be shipped?

**Comment by jsotuyod** [comment]
@filipesperandio usually, we ship once a month for minor releases. However, we are currently working on a major release, which is taking more time as we build all the breaking API changes.

We are closing in on the release (see [milestone status](https://github.com/pmd/pmd/milestone/12)) at 36 open issues / PRs, 102 closed.

It relies heavily on our time available to work on this. It has fluctuated a little recently, but I would love to have it shipped sooner rather than later.

**Comment by filipesperandio** [comment]
Cool! It is really just for my knowledge and curiosity.
It is nice to see this great tool under very active development.
Looking forward to the next release.

**Comment by jsotuyod** [comment]
Merged! I just added the unit test I mentioned. Thanks for the contribution!

**Pull request merged**